stdlib/public/runtime/DynamicCast.cpp

```diff
@@ -0,0 +1,2171 @@
+  if (isSuccess(subcastResult)) {
+    destInnerType->vw_storeEnumTagSinglePayload(
+      destLocation, /*case*/ 0, /*emptyCases*/ 1);
+  }
+  return subcastResult;
+}
+
+// Try unwrapping just the source optional.
+// Note we do this even if both target and dest are optional.
+// For example, this is used when extracting the contents of
+// an Optional<Any>.
+static DynamicCastResult
+tryCastUnwrappingOptionalSource(
+  OpaqueValue *destLocation, const Metadata *destType,
+  OpaqueValue *srcValue, const Metadata *srcType,
+  const Metadata *&destFailureType, const Metadata *&srcFailureType,
+  bool takeOnSuccess, bool mayDeferChecks)
+{
+  assert(srcType->getKind() == MetadataKind::Optional);
+
+  auto srcInnerType = cast<EnumMetadata>(srcType)->getGenericArgs()[0];
+  unsigned sourceEnumCase = srcInnerType->vw_getEnumTagSinglePayload(
+    srcValue, /*emptyCases=*/1);
+  auto nonNil = (sourceEnumCase == 0);
+  if (nonNil) {
+    // Recurse with unwrapped source
+    return tryCast(destLocation, destType, srcValue, srcInnerType,
+      destFailureType, srcFailureType, takeOnSuccess, mayDeferChecks);
+  }
+  return DynamicCastResult::Failure;
+}
+
+/******************************************************************************/
+/***************************** Tuple Destination ******************************/
+/******************************************************************************/
+
+// The only thing that can be legally cast to a tuple is another tuple.
+// Most of the logic below is required to handle element-wise casts of
+// tuples that are compatible but not of the same type.
+
+static DynamicCastResult
+tryCastToTuple(
+  OpaqueValue *destLocation, const Metadata *destType,
+  OpaqueValue *srcValue, const Metadata *srcType,
+  const Metadata *&destFailureType, const Metadata *&srcFailureType,
+  bool takeOnSuccess, bool mayDeferChecks)
+{
+  assert(srcType != destType);
+  assert(destType->getKind() == MetadataKind::Tuple);
+  const auto destTupleType = cast<TupleTypeMetadata>(destType);
+
+  srcFailureType = srcType;
+  destFailureType = destType;
+
+  // We cannot cast non-tuple data to a tuple
+  if (srcType->getKind() != MetadataKind::Tuple) {
+    return DynamicCastResult::Failure;
+  }
+  const auto srcTupleType = cast<TupleTypeMetadata>(srcType);
+
+  // Tuples must have same number of elements
+  if (srcTupleType->NumElements != destTupleType->NumElements) {
+    return DynamicCastResult::Failure;
+  }
+
+  // Common labels must match
+  const char *srcLabels = srcTupleType->Labels;
+  const char *destLabels = destTupleType->Labels;
+  while (srcLabels != nullptr
+         && destLabels != nullptr
+         && srcLabels != destLabels)
+  {
+    const char *srcSpace = strchr(srcLabels, ' ');
+    const char *destSpace = strchr(destLabels, ' ');
+
+    // If we've reached the end of either label string, we're done.
+    if (srcSpace == nullptr || destSpace == nullptr) {
```

Comment:
Is space used as a terminator? If it's just a separator then this will skip the last label.

Reply:
This matches the former logic as far as I can tell.

I expanded test/Interpreter/tuple_casts.swift to further probe tuple label checking and verified that it behaves identically with both old and new casting implementations.

Reply:
Is it detecting mismatches in the last label? If so, all good. If not, we have an unpleasant compatibility question to deal with, but I suspect that's not the case.

Reply:
I specifically added some checks to `test/Interpreter/tuple_casts.swift` to verify that we correctly handle mismatches in the last label, so I think we're good here.

Reply:
Alright, perfect!